ringbuffer.py

```diff
@@ -185,23 +189,52 @@ def try_write(self, data):
             self.array[position.index] = data
             self.writer.increment()
 
-    def _has_read_conflict(self, reader_position):
+    def _try_read_no_lock(self, reader: Pointer,
+                          length: int = 1) -> ctypes._SimpleCData:
+        position = reader.get()
         writer_position = self.writer.get()
-        return writer_position.counter <= reader_position.counter
 
-    def _try_read_no_lock(self, reader):
-        position = reader.get()
-        if self._has_read_conflict(position):
+        if writer_position.counter <= position.counter:
             if not self.active.value:
                 raise WriterFinishedError
             else:
                 raise WaitingForWriterError
 
-        data = self.array[position.index]
-        reader.increment()
-        return data
-
-    def try_read(self, reader):
+        max_length = writer_position.counter - position.counter
+
+        if max_length < length:
+            raise InsufficientDataError
+
+        remaining_buffer_length = self.slot_count - position.index
+        new_array = (self.c_type * length)()
+        if position.generation == writer_position.generation or \
+                remaining_buffer_length > length:
+            # If the writer is ahead of the reader in the current buffer
+            # generation, a simple memmove can copy the array.
+            ctypes.memmove(new_array,
+                           ctypes.addressof(self.array[position.index]),
+                           ctypes.sizeof(new_array))
+        else:
+            # When the writer has a greater generation count, two memmove
+            # calls are necessary to fill the new array with data.
+            first_slice_length = remaining_buffer_length
+            second_slice_length = length - first_slice_length
+            one_item_size = int(ctypes.sizeof(new_array) / length)
+
+            ctypes.memmove(new_array,
+                           ctypes.addressof(self.array[position.index]),
+                           first_slice_length * one_item_size)
+
+            if second_slice_length > 0:
+                ctypes.memmove(
+                    ctypes.addressof(new_array[first_slice_length]),
+                    ctypes.addressof(self.array[0]),
+                    second_slice_length * one_item_size)
```

Comment:
Nice catch! I think this logic can be simplified though:

```python
item_size = ctypes.sizeof(self.c_type)
remaining = self.slot_count - reader_position.index
ctypes.memmove(
    new_array,
    ctypes.addressof(self.array[reader_position.index]),
    item_size * min(length, remaining)
)
if length > remaining:
    ctypes.memmove(
        ctypes.addressof(new_array[remaining]),
        ctypes.addressof(self.array),
        (length - remaining) * item_size
    )

```

Reply:
Actually I went with the recommendation from Chris just at the time when you posted this.